src/hotspot/share/jeandle/jeandleAbstractInterpreter.cpp

```diff
@@ -2009,6 +2009,85 @@ void JeandleAbstractInterpreter::do_unified_newarray(Klass* array_klass) {
+{
+  int ndimensions = _bytecodes.get_dimensions();
+
+  bool will_link;
+  ciArrayKlass* array_klass = _bytecodes.get_klass(will_link)->as_array_klass();
+  assert(will_link, "multianewarray: not link");
+
+  // Note: Array classes are always initialized; no is_initialized check.
+
+  if (ndimensions == 1) {
+    // Use [a]newarray if only one dimension
+    Klass* klass = (Klass*)(array_klass->constant_encoding());
+    do_unified_newarray(klass);
+    return;
+  }
+
+  llvm::FunctionCallee callee = [&]() -> llvm::FunctionCallee {
+    switch (ndimensions) {
+      case 1:  ShouldNotReachHere(); break;
+      case 2:  return JeandleRuntimeRoutine::multianewarray2_callee(_module);
+      case 3:  return JeandleRuntimeRoutine::multianewarray3_callee(_module);
+      case 4:  return JeandleRuntimeRoutine::multianewarray4_callee(_module);
+      case 5:  return JeandleRuntimeRoutine::multianewarray5_callee(_module);
+      default: return JeandleRuntimeRoutine::multianewarrayN_callee(_module);
+    }
+  }();
+
+  llvm::PointerType* klass_type = llvm::PointerType::get(*_context, llvm::jeandle::AddrSpace::CHeapAddrSpace);
+  llvm::Value* array_klass_addr = _ir_builder.getInt64((intptr_t)(array_klass->constant_encoding()));
+  llvm::Value* array_klass_ptr = _ir_builder.CreateIntToPtr(array_klass_addr, klass_type);
+
+  llvm::CallInst* current_thread = call_java_op("jeandle.current_thread", {});
+
+  llvm::SmallVector<llvm::Value*> args;
+  args.push_back(array_klass_ptr);
+
+  if (ndimensions <= 5) {
+    ResourceMark rm;
+    // Get the lengths from the stack (first dimension is on top)
+    llvm::Value** length = NEW_RESOURCE_ARRAY(llvm::Value*, ndimensions);
+    for (int index = ndimensions - 1; index >= 0; index--) {
+      length[index] = _jvm->ipop();
+    }
+    // Reverse the dimension arguements
+    for (int index = 0; index < ndimensions; index++) {
+      args.push_back(length[index]);
+    }
+  } else {
+    // Create a java array for dimension sizes
+    Klass* int_array_klass = (Klass*)(ciTypeArrayKlass::make(T_INT)->constant_encoding());
+    llvm::Value* int_array_klass_addr = _ir_builder.getInt64((intptr_t)int_array_klass);
+    llvm::PointerType* klass_type = llvm::PointerType::get(*_context, llvm::jeandle::AddrSpace::CHeapAddrSpace);
+    llvm::Value* int_array_klass_ptr = _ir_builder.CreateIntToPtr(int_array_klass_addr, klass_type);
+
+    llvm::Value* dimensions_array_length = _ir_builder.getInt32(ndimensions);
+
+    llvm::CallInst* dimensions_array_oop = call_java_op("jeandle.newarray", {int_array_klass_ptr, dimensions_array_length});
+
+    // Fill-in it with values
+    for (int index = ndimensions - 1; index >= 0; index--) {
+      // No need to do boundary_check here
+      llvm::Value* array_base_offset = _ir_builder.getInt32(arrayOopDesc::base_offset_in_bytes(T_INT));
+      llvm::Value* array_base = _ir_builder.CreateInBoundsPtrAdd(dimensions_array_oop, array_base_offset, "array_element_base");
```

Comment:
```suggestion
      llvm::Value* array_base_offset = _ir_builder.getInt32(arrayOopDesc::base_offset_in_bytes(T_INT));
      llvm::Value* array_base = _ir_builder.CreateInBoundsPtrAdd(dimensions_array_oop, array_base_offset, "dimension_array_element_base");
```
Move out of this loop